### ISSUE TITLE
Fitting multiple models with resampled data

### ISSUE THREAD

**Issue by tchakravarty** [hadley/modelr#31]
_Cross-posted from [here](http://stackoverflow.com/q/40204405/1414455)._

@hadley More questions to try to understand the tidy data science model better. 

In the [tidy model of data science implemented in [`modelr`][2], resampled data are organized using [list-columns][3]: 

    library(modelr)
    library(tidyverse)
    
    # create the k-folds
    df_heights_resampled = heights %>% 
      crossv_kfold(k = 10, id = "Resample ID")

It is possible to `map` a model to each of the training datasets in the list-column `train` and to compute a performance metric by `map`ping onto the list-column `test`.

In case this needs to be done withe multiple models, this needs to be repeated for each of the models.

    # create a list of formulas 
    formulas_heights = formulas(
      .response = ~ income, 
      model1  = ~ height + weight + marital + sex,
      model2 = ~ height + weight + marital + sex + education
    ) 
    
    # fit each of the models in the list of formulas
    df_heights_resampled = df_heights_resampled %>% 
      mutate(
        model1 = map(train, function(train_data) {
          lm(formulas_heights[[1]], data = train_data)
        }),
        model2 = map(train, function(train_data) {
          lm(formulas_heights[[2]], data = train_data)
        })
      )
    
    # score the models on the test sets
    df_heights_resampled = df_heights_resampled %>% 
      mutate(
        rmse1 = map2_dbl(.x = model1, .y = test, .f = rmse),
        rmse2 = map2_dbl(.x = model2, .y = test, .f = rmse)
      )

which gives: 

    > df_heights_resampled
    # A tibble: 10 × 7
                train           test `Resample ID`   model1   model2    rmse1    rmse2
               <list>         <list>         <chr>   <list>   <list>    <dbl>    <dbl>
    1  <S3: resample> <S3: resample>            01 <S3: lm> <S3: lm> 58018.35 53903.99
    2  <S3: resample> <S3: resample>            02 <S3: lm> <S3: lm> 55117.37 50279.38
    3  <S3: resample> <S3: resample>            03 <S3: lm> <S3: lm> 49005.82 44613.93
    4  <S3: resample> <S3: resample>            04 <S3: lm> <S3: lm> 55437.07 51068.90
    5  <S3: resample> <S3: resample>            05 <S3: lm> <S3: lm> 48845.35 44673.88
    6  <S3: resample> <S3: resample>            06 <S3: lm> <S3: lm> 58226.69 54010.50
    7  <S3: resample> <S3: resample>            07 <S3: lm> <S3: lm> 56571.93 53322.41
    8  <S3: resample> <S3: resample>            08 <S3: lm> <S3: lm> 46084.82 42294.50
    9  <S3: resample> <S3: resample>            09 <S3: lm> <S3: lm> 59762.22 54814.55
    10 <S3: resample> <S3: resample>            10 <S3: lm> <S3: lm> 45328.48 41882.79

Question: 
=====
This can get cumbersome really fast if the number of models to be explored is large. `modelr` provides the `fit_with` function that allows to iterate over a number of models (as characterized by multiple formulae) but that does not seem to allow for a list-column like `train` in the model above. I am assuming that one of the `*map*` family of functions will make this possible (`invoke_map`?), but have not been able to figure out how. 

  [1]: http://r4ds.had.co.nz/
  [2]: https://github.com/hadley/modelr
  [3]: http://r4ds.had.co.nz/many-models.html#list-columns-1

**Comment by tchakravarty** [comment]
One solution is this: 

```
run_regression = function(train_data) {
  lapply(formulas_heights, function(formula) {
    lm(formula, data = train_data)
  })
}

df_heights_resampled %>% 
  mutate(
    `Models (list)` = map(.f = run_regression, .x = train)
  )
```

but then I do not seem to be able to `unnest` the `Models (list)` list-column. Now, having just read the manual, I can see that `unnest` is supposed to split into _rows_ and not columns, but then I am surprised that `unnest` worked in that fashion on an example I posted on SO [yesterday](http://stackoverflow.com/a/40196909/1414455).

Basically, the example was similar, in that I wanted to cycle over all the variables in a dataset (that matched a certain criterion) and compute the bivariate R2 with the response. This was a slight variation of the above requirement.

Here is the full code for that which seemed to me to be a good solution, and shows that `unnest` can indeed split out a list-column into individual _columns_ not _rows_: 

```{r}
library(tidyverse)
library(modelr)
library(lazyeval)
library(purrr)

# create the dataset
df_foo = wakefield::r_data_frame(
  n = 100,
  wakefield::r_series(wakefield::r_sample, j = 5, name = "cat"),
  Y = wakefield::normal()
)

make_r2_variable = function(var_name, train_data) {
  summary(stats::lm(Y ~ var_name, data = train_data))$r.squared
}

make_r2 = function(train_data) {
  summarise_at(
    .tbl =  data.frame(train_data),
    .cols = vars(starts_with("cat_")),
    .funs = funs(make_r2_variable(., train_data = train_data))
  )
  
}

df_foo = df_foo %>% 
  mutate_at(.funs = funs(as.factor), .cols = vars(starts_with("cat"))) %>% 
  crossv_kfold(k = 10, id = "id") %>% 
  mutate(
    R2 = map(.x = train, .f = make_r2)
  ) %>% 
  unnest(R2)
``` 

Obviously, the difference here is that `R2` is a `data.frame` with the _right_ orientation `1 x 5` (`Models (list)` is a `list`). Basically, what would be nice is to have a version of the `tidyr::separate` function that is able to take a list-column and remove one level of hierarchy mapping it to individual columns. I am not sure whether such a function already exists in `purrr` or `tidyr`, but would be happy to be directed to it. Alternatively, what would be an elegant way to work with list columns where we need to map the elements into other columns (list or otherwise)?
